server/auth_function/lambda_function.py

```diff
@@ -163,13 +163,32 @@ def populate_user_if_necessary(db_connection, event) -> None:
     else:
         user_name = event["request"]["userAttributes"]["custom:idp_username"]
 
+    cursor = db_connection.cursor()
+
+    # in the case of historical FAM user that has no user_guid stored, add their user_guid
+    # if user does not exist or user has a user_guid already, this update query will do nothing
+    query_add_guid= """
+        UPDATE app_fam.fam_user SET
+        user_guid={user_guid}
+        WHERE user_type_code={user_type_code}
+        and LOWER(user_name)={user_name}
+        and user_guid is null
+    """
+    sql_query_query_add_guid = sql.SQL(query_add_guid).format(
+        user_guid=sql.Literal(user_guid),
+        user_type_code=sql.Literal(user_type_code),
+        user_name=sql.Literal(user_name.lower()),
+    )
+    cursor.execute(sql_query_query_add_guid)
+
+    # insert new user, or update user information
     raw_query = """INSERT INTO app_fam.fam_user
         (user_type_code, user_guid, cognito_user_id, user_name, business_guid,
         create_user, create_date, update_user, update_date)
         VALUES( {user_type_code}, {user_guid}, {cognito_user_id}, {user_name}, {business_guid},
         CURRENT_USER, CURRENT_DATE, CURRENT_USER, CURRENT_DATE)
-        ON CONFLICT (user_type_code, lower(user_name)) DO
-        UPDATE SET user_guid = {user_guid},  cognito_user_id = {cognito_user_id}, business_guid = {business_guid};"""
+        ON CONFLICT (user_type_code, lower(user_guid)) DO
```

Comment:
you're right, I verified yesterday, in our production data, all the users that have user_guid are using all upper cases, I'll remove the lower here.

Reply:
Yes, the spec from IDIM seems pretty clear that the GUID values are all upper case. I would go with accepting the GUID values as supplied (don't do lower/upper), and compare them without lower/upper.

Reply:
I feel it doesn't hurt to add constraint by checking the lower case `CREATE UNIQUE INDEX fam_usr_uk ON app_fam.fam_user(user_type_code, lower(user_guid));`, just to avoid human errors if create user directly by calling api (unlikely to happen in prod, but could happen in dev or test).